src/core/expr/fexpr_cummax.cc

```diff
@@ -0,0 +1,113 @@
+//------------------------------------------------------------------------------
+// Copyright 2022 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "column/const.h"
+#include "column/cummax.h"
+#include "column/latent.h"
+#include "documentation.h"
+#include "expr/fexpr_func.h"
+#include "expr/eval_context.h"
+#include "expr/workframe.h"
+#include "python/xargs.h"
+#include "stype.h"
+
+
+namespace dt {
+namespace expr {
+
+class FExpr_cummax : public FExpr_Func {
+  private:
+    ptrExpr arg_;
+
+  public:
+    FExpr_cummax(ptrExpr&& arg)
+      : arg_(std::move(arg)) {}
+
+    std::string repr() const override{
+      std::string out = "cummax(";
+      out += arg_->repr();
+      out += ')';
+      return out;
+    }
+
+
+    Workframe evaluate_n(EvalContext& ctx) const override{
+      Workframe wf = arg_->evaluate_n(ctx);
+      Groupby gby = Groupby::single_group(wf.nrows());
+
+      if (ctx.has_groupby()) {
+        wf.increase_grouping_mode(Grouping::GtoALL);
+        gby = ctx.get_groupby();
+      }
+
+      for (size_t i = 0; i < wf.ncols(); ++i) {
+        Column coli = evaluate1(wf.retrieve_column(i), gby);
+        wf.replace_column(i, std::move(coli));
+      }
+      return wf;
+    }
+
+
+    Column evaluate1(Column&& col, const Groupby& gby) const {
+      SType stype = col.stype();
+      switch (stype) {
+        case SType::VOID:
+        case SType::BOOL:
+        case SType::INT8:
+        case SType::INT16:
+        case SType::INT32:
+        case SType::INT64: return make<int64_t>(std::move(col), SType::INT64, gby);
```

Comment:
`cummax()` should return the same type as the source column has. Here we return `int64` for all the integer/boolean/void types.

Reply:
Ahhh my bad... I'll fix it. I'll need ur help on templating and multiple functions. Couldn't wrap my head around the solution in rowminmax. I won't mind getting a solution for this with comments. I'll make the updates so you can input yours

Reply:
Sure, I can help you with templating this function. First, let's fix couple of issues in the current implementation of `cummax()`, then we will template it to also support `cummin()`.

Reply:
unsure what to use as the type for bool ... used bool8, but it seemed to have an issue with get_na<T>

Reply:
`bool` should be handled in the same way as `int8`.

Reply:
I assumed we needed to pass a make for book, similar to the other types

Reply:
Yes, but internally `bool` is equivalent to `int8`, so you just pass `int8` as a template parameter and the `stype` for the resulting column will be the same as the source column stype.